#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
NPU gRPC ä»£ç†æœåŠ¡ (Windows/Linux)

åŠŸèƒ½:
- æä¾› HTTP REST API æ¥å£
- è½¬å‘è¯·æ±‚åˆ°åä¸º NPU gRPC æœåŠ¡
- æ”¯æŒå›¾ç‰‡æ–‡ä»¶å¤¹æµè§ˆ
- æ”¯æŒä¸­æ–‡è·¯å¾„å’Œä¸­æ–‡æ ‡ç­¾

ä½¿ç”¨æ–¹æ³•:
    ç›´æ¥è¿è¡Œ: python npu_demo_server.py
    é…ç½®åœ¨ä¸‹æ–¹ CONFIG éƒ¨åˆ†ä¿®æ”¹
"""

import os
import sys
import base64
import time
import json
from pathlib import Path
from typing import List, Dict, Any, Optional, Tuple
from io import BytesIO

import cv2
import numpy as np
import grpc
from flask import Flask, request, jsonify, send_from_directory, send_file
from flask_cors import CORS

# ============================================================
# â˜…â˜…â˜… é…ç½®åŒºåŸŸ - åœ¨æ­¤å¤„ä¿®æ”¹é…ç½® â˜…â˜…â˜…
# ============================================================

# æœ¬åœ° HTTP æœåŠ¡é…ç½®
SERVER_HOST = '0.0.0.0'  # ç›‘å¬åœ°å€
SERVER_PORT = 8080  # æœ¬åœ°æœåŠ¡ç«¯å£

# è¿œç¨‹ NPU gRPC æœåŠ¡é…ç½®
GRPC_HOST = '172.18.8.11'  # NPU æœåŠ¡å™¨åœ°å€
GRPC_PORT = 8000  # NPU gRPC ç«¯å£
API_KEY = 'api-key'  # API å¯†é’¥

# å›¾ç‰‡ç›®å½•é…ç½®
IMAGE_DIR = './test_images'  # æµ‹è¯•å›¾ç‰‡ç›®å½•

# æ£€æµ‹å‚æ•°é»˜è®¤å€¼
DEFAULT_CONF = 0.25  # é»˜è®¤ç½®ä¿¡åº¦é˜ˆå€¼
DEFAULT_IOU = 0.45  # é»˜è®¤ IOU é˜ˆå€¼

# gRPC è¶…æ—¶æ—¶é—´ï¼ˆç§’ï¼‰
GRPC_TIMEOUT = 30


# ============================================================

# åŠ¨æ€ç”Ÿæˆ gRPC ä»£ç ï¼ˆå¦‚æœ proto æ–‡ä»¶å­˜åœ¨ï¼‰
def generate_grpc_code():
    """å°è¯•ä» proto æ–‡ä»¶ç”Ÿæˆ gRPC ä»£ç """
    proto_file = 'detection.proto'
    if os.path.exists(proto_file):
        try:
            import grpc_tools.protoc
            grpc_tools.protoc.main([
                'grpc_tools.protoc',
                f'--proto_path=.',
                f'--python_out=.',
                f'--grpc_python_out=.',
                proto_file
            ])
            print(f"[gRPC] å·²ä» {proto_file} ç”Ÿæˆä»£ç ")
        except Exception as e:
            print(f"[gRPC] ç”Ÿæˆä»£ç å¤±è´¥: {e}")


# å°è¯•å¯¼å…¥ gRPC ç”Ÿæˆçš„ä»£ç 
try:
    import detection_pb2
    import detection_pb2_grpc

    GRPC_AVAILABLE = True
except ImportError:
    GRPC_AVAILABLE = False
    print("[è­¦å‘Š] æœªæ‰¾åˆ° detection_pb2 æ¨¡å—ï¼Œå°†ä½¿ç”¨æ¨¡æ‹Ÿæ¨¡å¼")

# ============================================================
# ç®—æ³•é…ç½®ï¼ˆä¸ NPU æœåŠ¡ä¿æŒä¸€è‡´ï¼‰
# ============================================================
ALGO_CONFIG = {
    1: {"name": "æ¾çº¿è™«å®³è¯†åˆ«", "classes": ["æ­»äº¡", "é‡åº¦æ‚£ç—…", "è½»åº¦æ‚£ç—…"]},
    2: {"name": "æ°´æ±¡æŸ“æ£€æµ‹", "classes": ["æ°´æ±¡æŸ“", "æ¼‚æµ®ç¢ç‰‡", "åºŸå¼ƒèˆ¹åª", "æ¸”ä¸šå’Œæ°´äº§å…»æ®–", "åƒåœ¾"]},
    3: {"name": "æ°´é¢åƒåœ¾æ£€æµ‹", "classes": ["ç“¶å­", "è‰", "æ ‘æ", "ç‰›å¥¶ç›’", "å¡‘æ–™è¢‹", "å¡‘æ–™åƒåœ¾è¢‹", "çƒ", "å¶å­"]},
    4: {"name": "æ°´åŸŸå®‰å…¨æ£€æµ‹", "classes": ["å¿½ç•¥", "æ¸¸æ³³è€…", "èˆ¹", "æ°´ä¸Šæ‘©æ‰˜è‰‡", "æ•‘ç”Ÿè®¾å¤‡", "æµ®æ ‡"]},
    5: {"name": "è½¦ç‰Œè¯†åˆ«", "classes": ["è½¦ç‰Œ"]},
    6: {"name": "è½¦è¾†è¯†åˆ«", "classes": ["è½¦è¾†"]},
    7: {"name": "è·¯é¢ç—…å®³è¯†åˆ«",
        "classes": ["é¾Ÿè£‚", "çºµå‘è£‚ç¼", "çºµå‘ä¿®è¡¥å—", "æ£€æŸ¥äº•äº•ç›–", "å‘æ´", "æ¨ªå‘è£‚ç¼", "æ¨ªå‘ä¿®è¡¥å—"]},
    8: {"name": "åŸå¸‚éƒ¨ä»¶æ£€æµ‹",
        "classes": ["è¿è§„å¹¿å‘Šç‰Œ", "ç ´æŸæ ‡è¯†ç‰Œ", "äººè¡Œé“æ‚ç‰©å †ç§¯", "æ–½å·¥è·¯æ®µ", "è¤ªè‰²æ ‡è¯†ç‰Œ", "åƒåœ¾å †ç§¯", "æ¶‚é¸¦ä¹±ç”»",
                    "è·¯é¢å‘æ´", "è·¯é¢ç§¯æ²™", "å»ºç­‘å¤–ç«‹é¢ç ´æŸ"]},
    9: {"name": "äººè½¦æ£€æµ‹", "classes": ["è½¦", "äºº"]},
    10: {"name": "é˜²æººæ°´æ£€æµ‹", "classes": ["æ°´è¾¹é’“é±¼", "æ¸¸æ³³æººæ°´", "é’“é±¼ä¼", "èˆ¹"]},
    11: {"name": "å·¥ç¨‹æœºæ¢°æ£€æµ‹", "classes": ["èµ·é‡æœº", "æŒ–æ˜æœº", "æ‹–æ‹‰æœº", "å¡è½¦"]},
    12: {"name": "ç§¸ç§†ç„šçƒ§æ£€æµ‹", "classes": ["ç§¸ç§†å †"]},
    13: {"name": "å˜åŒ–æ£€æµ‹", "classes": ["æ— å˜åŒ–", "å˜åŒ–åŒºåŸŸ"]},
    14: {"name": "å é“ç»è¥è¯†åˆ«", "classes": ["å é“ç»è¥"]},
    15: {"name": "åŸå¸‚è¿è§„æ£€æµ‹",
         "classes": ["é•¿æ¤…", "å•†ä¸šåƒåœ¾", "éæ³•å€¾å€’ç‚¹", "ç»¿åœ°", "å­”æ´", "æ³½è¥¿æŠ¤æ ", "åœ°å—", "åŸææ–™", "ç”Ÿæ´»åƒåœ¾"]},
    16: {"name": "è£¸åœŸåƒåœ¾æ£€æµ‹", "classes": ["åƒåœ¾", "è£¸åœŸ"]},
    17: {"name": "è¿å»ºè¯†åˆ«", "classes": ["è“è‰²å¤©ç¯·", "å…¶ä»–è¿å»º", "æ”¹è£…ç»¿è‰²å°å±‹"]},
    18: {"name": "çƒŸç«æ£€æµ‹", "classes": ["çƒŸé›¾", "ç«"]},
    19: {"name": "å…‰ä¼æ¿æ£€æµ‹", "classes": ["æœ‰ç¼ºé™·çš„å…‰ä¼ç”µæ± "]},
    20: {"name": "è¡Œäººè½¦è¾†æ£€æµ‹", "classes": ["äºº", "è½¦", "è‡ªè¡Œè½¦"]},
    21: {"name": "å¢™ä½“ç—…å®³æ£€æµ‹", "classes": ["å¢™ä½“è…èš€", "å¢™ä½“å¼€è£‚", "å¢™ä½“åŠ£åŒ–", "å¢™æ¨¡", "å¢™é¢æ±¡æ¸"]},
    22: {"name": "ç½‚ç²Ÿè¯†åˆ«", "classes": ["ç½‚ç²Ÿ"]},
    23: {"name": "å†œä½œç‰©è¯†åˆ«", "classes": ["ä½œç‰©å€’ä¼"]},
    24: {"name": "è“è—»æ£€æµ‹", "classes": ["è“è—»"]},
    25: {"name": "èˆ¹åªæ£€æµ‹", "classes": ["èˆ¹åª"]},
}


# ============================================================
# gRPC å®¢æˆ·ç«¯
# ============================================================
class GRPCClient:
    """gRPC å®¢æˆ·ç«¯å°è£…"""

    def __init__(self, host: str, port: int, api_key: str):
        self.host = host
        self.port = port
        self.api_key = api_key
        self.channel = None
        self.stub = None

    def connect(self):
        """å»ºç«‹è¿æ¥"""
        if not GRPC_AVAILABLE:
            return False

        try:
            target = f"{self.host}:{self.port}"
            self.channel = grpc.insecure_channel(
                target,
                options=[
                    ('grpc.max_send_message_length', 50 * 1024 * 1024),
                    ('grpc.max_receive_message_length', 50 * 1024 * 1024),
                ]
            )
            self.stub = detection_pb2_grpc.DetectionServiceStub(self.channel)
            return True
        except Exception as e:
            print(f"[gRPC] è¿æ¥å¤±è´¥: {e}")
            return False

    def _get_metadata(self):
        """è·å–è¯·æ±‚å…ƒæ•°æ®"""
        return [('x-api-key', self.api_key)]

    def health_check(self) -> Dict:
        """å¥åº·æ£€æŸ¥"""
        if not self.stub:
            self.connect()

        try:
            request = detection_pb2.HealthRequest()
            response = self.stub.HealthCheck(
                request,
                metadata=self._get_metadata(),
                timeout=5
            )
            return {
                "status": response.status,
                "device": response.device,
                "models_cached": response.models_cached
            }
        except Exception as e:
            return {"status": "error", "message": str(e)}

    def get_version(self) -> Dict:
        """è·å–ç‰ˆæœ¬ä¿¡æ¯"""
        if not self.stub:
            self.connect()

        try:
            request = detection_pb2.VersionRequest()
            response = self.stub.GetVersion(
                request,
                metadata=self._get_metadata(),
                timeout=5
            )
            return {
                "version": response.version,
                "device": response.device,
                "algo_supported": list(response.algo_supported)
            }
        except Exception as e:
            return {"version": "unknown", "error": str(e)}

    def detect(self, image: np.ndarray, algorithm_id: int,
               conf_threshold: float = 0.25) -> Tuple[List[Dict], float]:
        """
        æ‰§è¡Œç›®æ ‡æ£€æµ‹

        Returns:
            (detections, detect_time)
        """
        if not self.stub:
            self.connect()

        # ç¼–ç å›¾ç‰‡ä¸º base64
        _, buffer = cv2.imencode('.jpg', image, [cv2.IMWRITE_JPEG_QUALITY, 95])
        image_b64 = base64.b64encode(buffer).decode('utf-8')

        # æ„å»ºè¯·æ±‚
        request = detection_pb2.DetectRequest(
            algorithm_id=algorithm_id,
            image=image_b64,
            conf_threshold=conf_threshold
        )

        # å‘é€è¯·æ±‚
        response = self.stub.Detect(
            request,
            metadata=self._get_metadata(),
            timeout=GRPC_TIMEOUT
        )

        if response.code != 200:
            raise ValueError(response.message)

        # è§£æç»“æœ
        detections = []
        for det in response.data.detections:
            detections.append({
                "class_id": det.class_id,
                "class_name": det.class_name,
                "confidence": det.confidence,
                "bbox": list(det.bbox)
            })

        return detections, response.data.detect_time

    def detect_change(self, image1: np.ndarray, image2: np.ndarray) -> Tuple[np.ndarray, float, float, int, int]:
        """
        æ‰§è¡Œå˜åŒ–æ£€æµ‹

        Args:
            image1: ç¬¬ä¸€å¼ å›¾ç‰‡ (BGR)
            image2: ç¬¬äºŒå¼ å›¾ç‰‡ (BGR)

        Returns:
            (mask_image, change_ratio, detect_time, width, height)
        """
        if not self.stub:
            self.connect()

        # ç¼–ç å›¾ç‰‡ä¸º base64 PNG
        _, buffer1 = cv2.imencode('.png', image1)
        _, buffer2 = cv2.imencode('.png', image2)
        image1_b64 = base64.b64encode(buffer1).decode('utf-8')
        image2_b64 = base64.b64encode(buffer2).decode('utf-8')

        # æ„å»ºè¯·æ±‚
        request = detection_pb2.ChangeDetectRequest(
            image1=image1_b64,
            image2=image2_b64
        )

        # å‘é€è¯·æ±‚
        response = self.stub.DetectChange(
            request,
            metadata=self._get_metadata(),
            timeout=GRPC_TIMEOUT
        )

        if response.code != 200:
            raise ValueError(response.message)

        # è§£ç  mask å›¾ç‰‡
        mask_data = base64.b64decode(response.mask)
        mask_array = np.frombuffer(mask_data, dtype=np.uint8)
        mask_image = cv2.imdecode(mask_array, cv2.IMREAD_GRAYSCALE)

        return mask_image, response.change_ratio, response.detect_time, response.width, response.height


# ============================================================
# å›¾ç‰‡æ–‡ä»¶å¤¹ç®¡ç†å™¨
# ============================================================
class ImageFolderManager:
    """å›¾ç‰‡æ–‡ä»¶å¤¹ç®¡ç†å™¨"""

    SUPPORTED_EXTENSIONS = {'.jpg', '.jpeg', '.png', '.bmp', '.gif', '.webp'}

    def __init__(self, image_dir: str):
        self.image_dir = image_dir

    def get_folders(self) -> List[Dict]:
        """è·å–æ‰€æœ‰å­æ–‡ä»¶å¤¹"""
        if not os.path.exists(self.image_dir):
            return []

        folders = []

        # æ£€æŸ¥å½“å‰ç›®å½•æ˜¯å¦æœ‰å›¾ç‰‡
        root_image_count = sum(1 for f in os.listdir(self.image_dir)
                               if os.path.isfile(os.path.join(self.image_dir, f)) and
                               Path(f).suffix.lower() in self.SUPPORTED_EXTENSIONS)
        if root_image_count > 0:
            folders.append({
                "name": "ğŸ“· å½“å‰ç›®å½•",
                "path": "__ROOT__",
                "image_count": root_image_count
            })

        # è·å–å­æ–‡ä»¶å¤¹
        for name in sorted(os.listdir(self.image_dir)):
            folder_path = os.path.join(self.image_dir, name)
            if os.path.isdir(folder_path):
                image_count = sum(1 for f in os.listdir(folder_path)
                                  if Path(f).suffix.lower() in self.SUPPORTED_EXTENSIONS)
                if image_count > 0:
                    folders.append({
                        "name": name,
                        "path": name,
                        "image_count": image_count
                    })

        return folders

    def get_images(self, folder_name: str) -> List[Dict]:
        """è·å–æ–‡ä»¶å¤¹ä¸­çš„å›¾ç‰‡åˆ—è¡¨"""
        if folder_name == "__ROOT__":
            folder_path = self.image_dir
        else:
            folder_path = os.path.join(self.image_dir, folder_name)

        if not os.path.exists(folder_path):
            return []

        images = []
        for filename in sorted(os.listdir(folder_path)):
            filepath = os.path.join(folder_path, filename)
            if os.path.isfile(filepath) and Path(filename).suffix.lower() in self.SUPPORTED_EXTENSIONS:
                size = os.path.getsize(filepath)
                if folder_name == "__ROOT__":
                    rel_path = filename
                else:
                    rel_path = f"{folder_name}/{filename}"
                images.append({
                    "name": filename,
                    "path": rel_path,
                    "size": size,
                    "size_str": self._format_size(size)
                })

        return images

    def get_image_path(self, relative_path: str) -> Optional[str]:
        """è·å–å›¾ç‰‡å®Œæ•´è·¯å¾„"""
        full_path = os.path.join(self.image_dir, relative_path)
        if os.path.exists(full_path):
            return full_path
        return None

    @staticmethod
    def _format_size(size: int) -> str:
        """æ ¼å¼åŒ–æ–‡ä»¶å¤§å°"""
        for unit in ['B', 'KB', 'MB', 'GB']:
            if size < 1024:
                return f"{size:.1f} {unit}"
            size /= 1024
        return f"{size:.1f} TB"


# ============================================================
# Flask åº”ç”¨
# ============================================================
app = Flask(__name__, static_folder=None)
CORS(app)

grpc_client: GRPCClient = None
image_manager: ImageFolderManager = None


@app.route('/')
def index():
    """è¿”å›å‰ç«¯é¡µé¢"""
    return send_from_directory('.', 'npu_demo.html')


@app.route('/api/config', methods=['POST', 'GET'])
def api_config():
    """è·å–æˆ–æ›´æ–°é…ç½®"""
    global image_manager, grpc_client

    if request.method == 'GET':
        return jsonify({
            "code": 200,
            "data": {
                "grpc_host": grpc_client.host if grpc_client else GRPC_HOST,
                "grpc_port": grpc_client.port if grpc_client else GRPC_PORT,
                "image_dir": image_manager.image_dir if image_manager else IMAGE_DIR
            }
        })

    # POST - æ›´æ–°é…ç½®
    try:
        data = request.get_json()

        # æ›´æ–° gRPC å®¢æˆ·ç«¯
        new_grpc_host = data.get('grpc_host')
        new_grpc_port = data.get('grpc_port')
        if new_grpc_host or new_grpc_port:
            host = new_grpc_host or grpc_client.host
            port = int(new_grpc_port) if new_grpc_port else grpc_client.port
            grpc_client = GRPCClient(host, port, API_KEY)
            grpc_client.connect()
            print(f"[é…ç½®] gRPC æœåŠ¡å·²æ›´æ–°: {host}:{port}")

        # æ›´æ–°å›¾ç‰‡ç®¡ç†å™¨
        new_image_dir = data.get('image_dir')
        if new_image_dir:
            if os.path.exists(new_image_dir):
                image_manager = ImageFolderManager(new_image_dir)
                print(f"[é…ç½®] å›¾ç‰‡ç›®å½•å·²æ›´æ–°: {new_image_dir}")
            else:
                return jsonify({"code": 400, "message": f"å›¾ç‰‡ç›®å½•ä¸å­˜åœ¨: {new_image_dir}"})

        return jsonify({
            "code": 200,
            "message": "é…ç½®å·²æ›´æ–°"
        })
    except Exception as e:
        return jsonify({"code": 500, "message": str(e)})


@app.route('/api/health')
def api_health():
    """å¥åº·æ£€æŸ¥"""
    result = grpc_client.health_check()
    return jsonify({"code": 200, "data": result})


@app.route('/api/algorithms')
def get_algorithms():
    """è·å–æ”¯æŒçš„ç®—æ³•åˆ—è¡¨"""
    algorithms = []
    for algo_id, config in sorted(ALGO_CONFIG.items()):
        algorithms.append({
            "id": algo_id,
            "name": config["name"],
            "classes": config["classes"],
            "type": "change_detection" if algo_id == 13 else "detection"
        })
    return jsonify({"code": 200, "data": algorithms})


@app.route('/api/folders')
def get_folders():
    """è·å–å›¾ç‰‡æ–‡ä»¶å¤¹åˆ—è¡¨"""
    return jsonify({
        "code": 200,
        "data": image_manager.get_folders()
    })


@app.route('/api/images/<path:folder_name>')
def get_images(folder_name: str):
    """è·å–æ–‡ä»¶å¤¹ä¸­çš„å›¾ç‰‡åˆ—è¡¨"""
    return jsonify({
        "code": 200,
        "data": image_manager.get_images(folder_name)
    })


@app.route('/api/image/<path:image_path>')
def get_image(image_path: str):
    """è·å–å›¾ç‰‡æ–‡ä»¶"""
    full_path = image_manager.get_image_path(image_path)
    if full_path:
        return send_file(full_path)
    return jsonify({"code": 404, "message": "å›¾ç‰‡ä¸å­˜åœ¨"}), 404


@app.route('/api/thumb/<path:image_path>')
def get_thumbnail(image_path: str):
    """è·å–å›¾ç‰‡ç¼©ç•¥å›¾"""
    full_path = image_manager.get_image_path(image_path)
    if not full_path:
        return jsonify({"code": 404, "message": "å›¾ç‰‡ä¸å­˜åœ¨"}), 404

    try:
        # è¯»å–å›¾ç‰‡ï¼ˆæ”¯æŒä¸­æ–‡è·¯å¾„ï¼‰
        image = cv2.imdecode(np.fromfile(full_path, dtype=np.uint8), cv2.IMREAD_COLOR)
        if image is None:
            return jsonify({"code": 400, "message": "æ— æ³•è¯»å–å›¾ç‰‡"}), 400

        # ç”Ÿæˆç¼©ç•¥å›¾
        h, w = image.shape[:2]
        max_size = 100
        if w > h:
            new_w = max_size
            new_h = int(h * max_size / w)
        else:
            new_h = max_size
            new_w = int(w * max_size / h)

        thumb = cv2.resize(image, (new_w, new_h), interpolation=cv2.INTER_AREA)
        _, buffer = cv2.imencode('.jpg', thumb, [cv2.IMWRITE_JPEG_QUALITY, 70])

        return send_file(BytesIO(buffer.tobytes()), mimetype='image/jpeg')
    except Exception as e:
        return jsonify({"code": 500, "message": str(e)}), 500


@app.route('/api/detect', methods=['POST'])
def detect():
    """æ‰§è¡Œç›®æ ‡æ£€æµ‹"""
    data = request.json

    algo_id = data.get('algorithm_id')
    image_path = data.get('image_path')
    conf_threshold = data.get('conf_threshold', DEFAULT_CONF)

    if not algo_id:
        return jsonify({"code": 400, "message": "ç¼ºå°‘ algorithm_id"}), 400

    if not image_path:
        return jsonify({"code": 400, "message": "ç¼ºå°‘ image_path"}), 400

    # è·å–å›¾ç‰‡
    full_path = image_manager.get_image_path(image_path)
    if not full_path:
        return jsonify({"code": 404, "message": "å›¾ç‰‡ä¸å­˜åœ¨"}), 404

    # è¯»å–å›¾ç‰‡ï¼ˆæ”¯æŒä¸­æ–‡è·¯å¾„ï¼‰
    try:
        image = cv2.imdecode(np.fromfile(full_path, dtype=np.uint8), cv2.IMREAD_COLOR)
    except Exception as e:
        return jsonify({"code": 400, "message": f"è¯»å–å›¾ç‰‡å¤±è´¥: {e}"}), 400

    if image is None:
        return jsonify({"code": 400, "message": "æ— æ³•è¯»å–å›¾ç‰‡"}), 400

    # è°ƒç”¨ gRPC æœåŠ¡æ£€æµ‹
    try:
        detections, detect_time = grpc_client.detect(image, algo_id, conf_threshold)
    except Exception as e:
        return jsonify({"code": 500, "message": f"æ£€æµ‹å¤±è´¥: {e}"}), 500

    # ç»˜åˆ¶æ£€æµ‹ç»“æœ
    result_img = draw_detections(image, detections)

    # ç¼–ç ä¸º base64
    _, buffer = cv2.imencode('.jpg', result_img, [cv2.IMWRITE_JPEG_QUALITY, 90])
    result_b64 = base64.b64encode(buffer).decode('utf-8')

    # è·å–ç®—æ³•åç§°
    algo_name = ALGO_CONFIG.get(algo_id, {}).get("name", f"ç®—æ³•{algo_id}")

    return jsonify({
        "code": 200,
        "data": {
            "algorithm_id": algo_id,
            "algorithm_name": algo_name,
            "detections": detections,
            "total_count": len(detections),
            "detect_time": detect_time,
            "result_image": f"data:image/jpeg;base64,{result_b64}"
        }
    })


@app.route('/api/detect_change', methods=['POST'])
def detect_change():
    """æ‰§è¡Œå˜åŒ–æ£€æµ‹ï¼ˆéœ€è¦ä¸¤å¼ å›¾ç‰‡ï¼‰"""
    data = request.json

    image_path1 = data.get('image_path1')
    image_path2 = data.get('image_path2')

    if not image_path1 or not image_path2:
        return jsonify({"code": 400, "message": "éœ€è¦æä¾›ä¸¤å¼ å›¾ç‰‡è·¯å¾„ (image_path1, image_path2)"}), 400

    # è·å–å›¾ç‰‡1
    full_path1 = image_manager.get_image_path(image_path1)
    if not full_path1:
        return jsonify({"code": 404, "message": f"å›¾ç‰‡1ä¸å­˜åœ¨: {image_path1}"}), 404

    # è·å–å›¾ç‰‡2
    full_path2 = image_manager.get_image_path(image_path2)
    if not full_path2:
        return jsonify({"code": 404, "message": f"å›¾ç‰‡2ä¸å­˜åœ¨: {image_path2}"}), 404

    # è¯»å–å›¾ç‰‡
    try:
        image1_orig = cv2.imdecode(np.fromfile(full_path1, dtype=np.uint8), cv2.IMREAD_COLOR)
        image2_orig = cv2.imdecode(np.fromfile(full_path2, dtype=np.uint8), cv2.IMREAD_COLOR)
    except Exception as e:
        return jsonify({"code": 400, "message": f"è¯»å–å›¾ç‰‡å¤±è´¥: {e}"}), 400

    if image1_orig is None or image2_orig is None:
        return jsonify({"code": 400, "message": "æ— æ³•è¯»å–å›¾ç‰‡"}), 400

    # å˜åŒ–æ£€æµ‹éœ€è¦ç»Ÿä¸€å°ºå¯¸åˆ° 256x256
    input_size = 256
    image1 = cv2.resize(image1_orig, (input_size, input_size))
    image2 = cv2.resize(image2_orig, (input_size, input_size))

    # è°ƒç”¨ gRPC å˜åŒ–æ£€æµ‹æœåŠ¡
    try:
        mask_gray, change_ratio, detect_time, width, height = grpc_client.detect_change(image1, image2)
    except Exception as e:
        return jsonify({"code": 500, "message": f"æ£€æµ‹å¤±è´¥: {e}"}), 500

    # åˆ›å»ºå½©è‰² mask (çº¢è‰²è¡¨ç¤ºå˜åŒ–)
    mask_color = np.zeros((input_size, input_size, 3), dtype=np.uint8)
    if mask_gray is not None:
        mask_color[mask_gray > 127] = [0, 0, 255]  # BGR: çº¢è‰²

    # åˆ›å»ºå åŠ ç»“æœå›¾
    overlay = image2.copy()
    overlay[mask_gray > 127] = [0, 0, 255]
    blended = cv2.addWeighted(image2, 0.7, overlay, 0.3, 0)

    # ç¼–ç ç»“æœå›¾
    _, buffer1 = cv2.imencode('.jpg', image1, [cv2.IMWRITE_JPEG_QUALITY, 90])
    _, buffer2 = cv2.imencode('.jpg', image2, [cv2.IMWRITE_JPEG_QUALITY, 90])
    _, buffer_mask = cv2.imencode('.jpg', mask_color, [cv2.IMWRITE_JPEG_QUALITY, 90])
    _, buffer_blend = cv2.imencode('.jpg', blended, [cv2.IMWRITE_JPEG_QUALITY, 90])

    image1_b64 = base64.b64encode(buffer1).decode('utf-8')
    image2_b64 = base64.b64encode(buffer2).decode('utf-8')
    mask_b64 = base64.b64encode(buffer_mask).decode('utf-8')
    blend_b64 = base64.b64encode(buffer_blend).decode('utf-8')

    return jsonify({
        "code": 200,
        "data": {
            "algorithm_id": 13,
            "algorithm_name": "å˜åŒ–æ£€æµ‹",
            "image_width": input_size,
            "image_height": input_size,
            "change_ratio": round(change_ratio * 100, 2),
            "detect_time": round(detect_time, 3),
            "image1": f"data:image/jpeg;base64,{image1_b64}",
            "image2": f"data:image/jpeg;base64,{image2_b64}",
            "mask_image": f"data:image/jpeg;base64,{mask_b64}",
            "result_image": f"data:image/jpeg;base64,{blend_b64}"
        }
    })


def draw_detections(image: np.ndarray, detections: List[Dict]) -> np.ndarray:
    """åœ¨å›¾ç‰‡ä¸Šç»˜åˆ¶æ£€æµ‹ç»“æœï¼ˆæ”¯æŒä¸­æ–‡ï¼‰"""
    from PIL import Image, ImageDraw, ImageFont

    # è½¬æ¢ä¸º PIL å›¾åƒ
    result = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
    pil_image = Image.fromarray(result)
    draw = ImageDraw.Draw(pil_image)

    # å°è¯•åŠ è½½ä¸­æ–‡å­—ä½“
    font = None
    font_size = 20
    font_paths = [
        "C:/Windows/Fonts/msyh.ttc",
        "C:/Windows/Fonts/simsun.ttc",
        "C:/Windows/Fonts/simhei.ttf",
        "/usr/share/fonts/truetype/droid/DroidSansFallbackFull.ttf",
        "/System/Library/Fonts/PingFang.ttc",
    ]
    for font_path in font_paths:
        if os.path.exists(font_path):
            try:
                font = ImageFont.truetype(font_path, font_size)
                break
            except:
                continue

    if font is None:
        try:
            font = ImageFont.truetype("arial.ttf", font_size)
        except:
            font = ImageFont.load_default()

    # é¢œè‰²åˆ—è¡¨ (RGB)
    colors = [
        (0, 255, 0), (255, 0, 0), (0, 0, 255), (255, 255, 0),
        (255, 0, 255), (0, 255, 255), (128, 255, 0), (255, 128, 0),
        (128, 0, 255), (0, 128, 255), (255, 0, 128), (0, 255, 128),
    ]

    for det in detections:
        bbox = det["bbox"]
        x1, y1, x2, y2 = map(int, bbox)
        class_id = det["class_id"]
        class_name = det["class_name"]
        confidence = det["confidence"]

        color = colors[class_id % len(colors)]

        # ç»˜åˆ¶è¾¹ç•Œæ¡†
        draw.rectangle([x1, y1, x2, y2], outline=color, width=3)

        # ç»˜åˆ¶æ ‡ç­¾
        label = f"{class_name} {confidence:.2f}"

        try:
            bbox_text = draw.textbbox((0, 0), label, font=font)
            text_w = bbox_text[2] - bbox_text[0]
            text_h = bbox_text[3] - bbox_text[1]
        except:
            text_w, text_h = len(label) * 10, 20

        label_y = max(y1 - text_h - 8, 0)
        draw.rectangle([x1, label_y, x1 + text_w + 10, label_y + text_h + 6], fill=color)
        draw.text((x1 + 5, label_y + 2), label, fill=(255, 255, 255), font=font)

    # è½¬æ¢å› OpenCV æ ¼å¼
    result = cv2.cvtColor(np.array(pil_image), cv2.COLOR_RGB2BGR)
    return result


def main():
    global grpc_client, image_manager

    print("=" * 60)
    print("NPU gRPC ä»£ç†æœåŠ¡")
    print("=" * 60)
    print(f"æœ¬åœ°æœåŠ¡: http://{SERVER_HOST}:{SERVER_PORT}")
    print(f"è¿œç¨‹ NPU: {GRPC_HOST}:{GRPC_PORT}")
    print(f"å›¾ç‰‡ç›®å½•: {IMAGE_DIR}")
    print("=" * 60)

    # æ£€æŸ¥ç›®å½•
    if not os.path.exists(IMAGE_DIR):
        print(f"\n[è­¦å‘Š] å›¾ç‰‡ç›®å½•ä¸å­˜åœ¨ï¼Œæ­£åœ¨åˆ›å»º: {IMAGE_DIR}")
        os.makedirs(IMAGE_DIR, exist_ok=True)

    # åˆå§‹åŒ–
    grpc_client = GRPCClient(GRPC_HOST, GRPC_PORT, API_KEY)
    image_manager = ImageFolderManager(IMAGE_DIR)

    # å°è¯•è¿æ¥ gRPC æœåŠ¡
    if GRPC_AVAILABLE:
        if grpc_client.connect():
            health = grpc_client.health_check()
            if health.get("status") == "ok":
                print(f"\nâœ… å·²è¿æ¥åˆ° NPU æœåŠ¡: {health.get('device')}")
            else:
                print(f"\nâš ï¸ NPU æœåŠ¡è¿æ¥å¼‚å¸¸: {health}")
        else:
            print("\nâš ï¸ æ— æ³•è¿æ¥åˆ° NPU æœåŠ¡ï¼Œè¯·æ£€æŸ¥ç½‘ç»œ")
    else:
        print("\nâš ï¸ gRPC æ¨¡å—æœªåŠ è½½ï¼Œè¯·ç¡®ä¿ detection_pb2.py å­˜åœ¨")

    # å¯åŠ¨æœåŠ¡
    print(f"\nğŸš€ æœåŠ¡å·²å¯åŠ¨: http://localhost:{SERVER_PORT}")
    print(f"   æµè§ˆå™¨è®¿é—®ä¸Šè¿°åœ°å€å³å¯ä½¿ç”¨\n")
    app.run(host=SERVER_HOST, port=SERVER_PORT, debug=False, threaded=True)


if __name__ == '__main__':
    main()